### DISCUSSION TITLE
How to install anyrun using Nix Home Manager?

### DISCUSSION THREAD

**Discussion by s1syph0s** [anyrun-org/anyrun#186] [Q&A] [answered]
Hey, I'm trying to install anyrun using home manager. Since I can't find the options in the home manager repo, I use it through as in input in my flake.nix.

`flake.nix`:
```
# ...
      homeConfigurations = {
        "sisyph0s@saturn" = home-manager.lib.homeManagerConfiguration {
	  inherit pkgs;
	  modules = [ 
	    ./host/saturn/home.nix 
	    {
	      inherit anyrun;
	    }
	  ];
	};
```

`home.nix`:
```
{ config, pkgs, anyrun, ... }:

{
  # Home Manager needs a bit of information about you and the paths it should
  # manage.
  home.username = "sisyph0s";
  home.homeDirectory = "/home/sisyph0s";

  fonts.fontconfig.enable = true;

  imports = [
    (import ../../common { inherit config pkgs anyrun; })
  ];
  # ...
```

`common/default.nix`:
```
{ config, pkgs, lib, anyrun, ... }:
{
  programs.anyrun = {
    enable = true;
    # ...
  };
}
```

but i got the following error:
```
error:
       … while evaluating a branch condition

         at /nix/store/y6205wq8hxvpqvl8l9d1n9xah01kg0lq-source/lib/lists.nix:125:9:

          124|       fold' = n:
          125|         if n == len
             |         ^
          126|         then nul

       … while calling the 'length' builtin

         at /nix/store/y6205wq8hxvpqvl8l9d1n9xah01kg0lq-source/lib/lists.nix:123:13:

          122|     let
          123|       len = length list;
             |             ^
          124|       fold' = n:

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called without required argument 'lib'

       at /nix/store/z37b9scin9dqq3nz6mbxd63px8bzna24-source/common/default.nix:1:1:

            1| { config, pkgs, lib, anyrun, ... }:
             | ^
            2| {
```

Where did I make a mistake? I feel like I'm still pretty new in nix and I'm sure that I made a mistake. Thanks for the help!

**Comment by haennes** [comment, marked as answer]
To pass the anyrun argument to your home-manager modules use extraSpecialArgs like demonstrated here: 
https://nix-community.github.io/home-manager/
for the normal module system look [here](https://nixos.org/manual/nixos/stable/options)

In your case this would be:
```
# ...
      homeConfigurations = {
        "sisyph0s@saturn" = home-manager.lib.homeManagerConfiguration {
	  inherit pkgs;
         extraSpecialArgs = {
              inherit anyrun;
         };
	  modules = [ 
	    ./host/saturn/home.nix 
	  ];
	};
```

**BUT this is probably not what you want.**
i assume your flake looks something like this:
```
...
inputs.anyrun.url = "https://github.com/anyrun-org/anyrun";
....
outputs = inputs@{..., anyrun, ...}:{
....
}
```
then you should just change the section to
```
# ...
      homeConfigurations = {
        "sisyph0s@saturn" = home-manager.lib.homeManagerConfiguration {
	  inherit pkgs;
	  modules = [ 
	    ./host/saturn/home.nix 
	    anyrun.nixosModules.anyrun
	  ];
	};
```
adding the module provides you with the options to enable anyrun, in any home-manager module

this should then be converted to:
```
{ config, pkgs,  ... }:

{
  # Home Manager needs a bit of information about you and the paths it should
  # manage.
  home.username = "sisyph0s";
  home.homeDirectory = "/home/sisyph0s";

  fonts.fontconfig.enable = true;

  imports = [
    ../../common
  ];
  # ...
```
and remove the anyrun argument in the home.nix


Now on the error that is thrown:
In common/default.nix your function take as lib argument, but when you import it in home.nix you didn´t pass the lib argument (could just be added to the inherit). However importing it the way you tried to has the problem that home-manager passes a couple of inputs to the modules (and imports, as they are just modules) and you would have to write them for each module you are importing.

Additional note, you probably want to use the home-manager module to in your configuration for your host, so applying home-manager and nixos changes is both done on `nixos-rebuild switch --flake .#`

**Reply by s1syph0s** [reply]
Thank you!
